Extend docstrings for amplitude scaling and collisions

JoeZiminski opened the pull request:
In investigating #2882 I had some trouble understanding what the code around amplitude scalings and fit collisions. The code itself is easy to follow, but there is a lack of context and supporting information on data structures which can make it tricky to understand what is going on without breaking into the code. I think I'll add the fix for #2882 in a different PR so the scope does not become too large.

This PR:
1) does some minor variable renamings for clarity
2) extends docstrings for `find_collisions`, `fit_collision` and `AmplitudeScaling` classes and adds a few extra explanatory comments.
3) Uncomments the plot-debugging code, makes functions private and add short docstring.

Some questions:
1) A lot of the computations depend on the first entry into the `collisions` structure being the spike of interest. If the code is refactored in a way that changes this by accident, it will be difficult to track down but have major negative affect (all amplitude scalings will be wrong). Make the structure of `collisions` can be changed, so this has two entries (`spike`) and (`colliding_spikes`). This imo would make this more robust in future.
2) The `spikes` structure is an array with entries `An array of spikes (sample_index, channel_index, amplitude, segment_index, unit_index, in_margin)`. I had added this explicitly to docstrings (initially, I just assumed it was array of indexes). But, there is now some duplication in the docstring, personally I don't think this too bad a thing, but might be a point for discussion.

Update before merge
- tests were not adressed here but can be later, based on [this comment](https://github.com/SpikeInterface/spikeinterface/pull/2893#discussion_r1610442163)
- the implicit spike-of-interest required to be in the first position of the collisions (1 above) dict is quite fiddly to address, and it would be easy to design a test to catch this case, this can be open in a new PR.

zm711 commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -13,16 +13,24 @@
+    has an associated amplitude scaling. This is a more robust
+    estimate of the waveform amplitude that other methods e.g. raw voltage.
+
+    When there are collisions, the scaling is the regression
+    of the waveform onto the template, with intercept:
+        scaling * template + intercept = waveform
+    When there are collisions, a multiple-regression is used to fit
+    the waveform to template while taking into account colliding spikes.
```
This could probably be simplified a little no? Seems like you're repeating a bit that it is the regression then you say it is a multiple-regression.

zm711 commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -383,28 +395,38 @@ def _are_unit_indices_overlapping(sparsity_mask, i, j):
+    First for each spike, find all other spikes that temporally overlap the spike.
+    Temporal overlap is defined as another spike peak occurring within a time window around
+    the spike peak. Next, only these temporally overlapping spikes that also
+    spatially overlap the spike are included in the output `collision_spikes_dict`.
```
temporal overlap is defined, but not spatial.

JoeZiminski replied:
Thanks I changed to
```
    Spatial overlap is defined as the two spikes both having signal on any shared
    channel (i.e. the spikes are not spatially overlapping if they have signal on
    two disjoint sets of channels).
```

zm711 commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -525,133 +576,128 @@ def fit_collision(
+    ----------
+
+    Parameters
+    we : SortingAnalyzer
```
should we switch this over to sorting_analyzer as an argument? I think leaving this as `we` is leaving some tech debt.

JoeZiminski replied:
I have switche to `sorting_analyzer`, but based on below comment I think this will break still.

zm711 commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -525,133 +576,128 @@ def fit_collision(
+    num_collisions : int, default=None
+        Number of collisions to plot. If None, all collisions are plotted.
+    """
+    assert we.is_extension("amplitude_scalings"), "Could not find amplitude scalings extension!"
+    sac = we.load_extension("amplitude_scalings")
+    handle_collisions = sac._params["handle_collisions"]
+    assert handle_collisions, "Amplitude scalings was run without handling collisions!"
+    scalings = sac.get_data()
+
```
yeah this is old so probably need to convert to sorting_analyzer. (is_extension for example was already deprecated even in 100.x)

JoeZiminski replied:
That sounds good, maybe this PR is a good opportunity to add tests for these plotting functions (to check that the conversion to sorting_analyzer is correct) and also extend the tests for `amplitude_scaling.py`.  I am not ATM familiar with the `AnalyzerExtensioncommonTestSuite` (looks very nice tho BTW) but as far as I can tell this checks the pipeline node runs without error, with and without `handle_collisions` [here](https://github.com/JoeZiminski/spikeinterface/blob/d37c2bc646180115f9d7108a7dfedde78ab847a5/src/spikeinterface/postprocessing/tests/test_amplitude_scalings.py#L19). And without handling collisions, it checks that the median scalings are some reasonable value. I'd be happy to extend this to:

1) unit test fit and find collisions, also `_units_are_spatially_overlapping` with some generated test input against expected output
2) perform a quick smoke test on `_plot_one_collision`, `_plot_collisions` to check they run without crashing. They are only internal for debugging anyways.
3) Testing `ComputeAmplitudeScalings` outputs are as expected. I am not certain on how best to do this. Maybe something like running it with toy data, then inspecting the scalings (under different input arguments) for some random spikes to check indeed the scaling is (exactly? or close to) the linear regression of waveform onto template. I think it will be hard to test this in the case of collisions, but at least the unit tests will go reasonably far for that.

zm711 reviewed (commented):
Just a few initial comments.

alejoe91 commented:
Thanks Joe! I'd be happy to discuss about this during the hackahton.

I brief comment about the `_plot_*` tests. These are designed as debug plotting routines, so I don't think they should be tested here. If we want to make a proper plotting routine for collisions (which might be cool for testing sorters and debugging), it's a better idea to port them to the `widgets` module

h-mayorquin reviewed (commented):
Thanks a lot for doing this. I learn from these efforts as I have not used the methods myself.

zm711 commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -367,14 +382,16 @@ def get_trace_margin(self):
 
 ### Collision handling ###
-def _are_unit_indices_overlapping(sparsity_mask, i, j):
+def _units_are_spatially_overlapping(sparsity_mask, i, j):
```
I think the difference is:

is the function answering a question (start with is/are)
is the function stating a fact (start with subject then do is/are)

In general I think this codebase tends toward wording these as questions rather than as statements.

JoeZiminski replied:
Interesting yes, I usually have thought about this in the context of how they read in conditional statements. For example:

```
if units_are_overlapping():
    ....
```
reads more naturally than:
```
if are_units_overlapping():
    ...
```

but just looking at it from the perspective of a function name, `are_units_overlapping()` gives more indication of what the function does (answering a question) than `units_are_overlapping()` (almost requires already knowing this returns a boolean to realise it is answering a question). So I'm not so strong on either approach and having a quick look
[1](https://stackoverflow.com/questions/1566745/boolean-method-naming-readability)
[2](https://stackoverflow.com/questions/47996388/python-boolean-methods-naming-convention)
[3](https://softwareengineering.stackexchange.com/questions/221868/should-methods-that-return-boolean-be-named-after-a-question-or-an-assertion)
. But maybe now is a good time to add something on this to the contributing guide or a style guide?

alejoe91 replied:
I vote fot the question style: `are_units_overlapping` (?)

zm711 replied:
I vote for the question as well. It's what the library is already doing so I think changing this midstream would be a big refactor. @JoeZiminski I would say feel free to add a note to the style guide!

h-mayorquin replied:
Makes sense to keep the same style of questions for booleans then. Questions it is.

JoeZiminski replied:
thanks all! I have reverted to question format

h-mayorquin commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -415,33 +445,43 @@ def find_collisions(spikes, spikes_w_margin, delta_collision_samples, sparsity_m
 
-        # find the possible spikes per and post within delta_collision_samples
+        # find the index of the spike within spikes_within_margin
+        spike_index_within_margin = np.where(spikes_within_margin == spike)[0][0]
```
I think you are right. My coment was that I like the semantics of `np.where` a bit more. As in, it is more clear what it is doing. But now that I think about it as I write this, maybe not so clear x D

JoeZiminski replied:
Cool, I did not know about nonzero! I see indeed numpy does [recommend its use](https://numpy.org/doc/stable/reference/generated/numpy.where.html) when the additional arguments are not provided. I don't quite understand the reason, they say `Using nonzero directly should be preferred, as it behaves correctly for subclasses.` but I'm not sure what that means - if `np.where` is just wrapping `np.nonzero` in this case then I wonder why it is performing differently? 

Personally I prefer `np.where` because:
1) It reads like a normal sentence "spike_index_within_margin equals where 'spikes_within_margin' equals  'spike'"
2) If mixing `np.nonzero` and `np.where` there are now two functions in use for doing essentially the same job.

Maybe we can make an issue on this or have some discussion it on a contributing / a style guide? If it's okay with everyone I will leave it as-is in this PR as it wasn't a change I introduced specifically but the additional was from refactoring some already-existing code.

zm711 replied:
Yeah barring some major issue I think it's fine. An issue to discuss this sounds fine.

h-mayorquin replied:
Given that the three of us prefer np.where instead of np.nonzero and there is not a performance cost maybe let's just adapta that convention. I honestly don't think that Sam or Alessio care about these things : )

![image](https://github.com/SpikeInterface/spikeinterface/assets/6429509/9073e723-6a99-430e-a0ff-7f0ff1b1a054)

JoeZiminski replied:
Nice! that sounds good to me, I will open a PR to make some additions to the style guide

h-mayorquin commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -457,29 +497,37 @@ def fit_collision(
+    `X` is the corresponding set of unit templates, with each template
+    temporally shifted to match the position of its associated spike in the
+    original waveform. The regression coefficients represent the scaling
+    factors applied to each template to best match the waveform.
```
Mmm that's a good point. Not sure. Good idea to discuss it on the docstrings hackaton discussion : )

JoeZiminski replied:
Yes I agree this would be great to discuss it is a very interesting feature. Maybe we could have a general issue on style-related questions for posting discussions such as the the ones raised in the PR until we get around to addressing them?

For me the latex-styling is really cool but makes the docstring itself hard to read, whereas not having the latex makes the docs look less nice but at least the docstrings are still easily readable. So on the possibilities for (docstring, docs) of (hard to read, awesome to read) vs (okay to read, okay to read) I tend to the latter.

zm711 replied:
I agree with Joe on both points. Unless we move over to do docstring testing/other crazy doctoring format I would prefer to be okay on both rather than bad on one and good on the other.

h-mayorquin replied:
For my own bias I rather make the docstring readable as a developer. Most of the time I am reading the code on an editor and not on a webpage.

JoeZiminski replied:
Thanks both I'll leave this for now then

JoeZiminski commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -13,16 +13,29 @@
 
+    Amplitude scalings are the scaling factor to multiply the
+    unit template to best match the waveform. Each waveform
+    has an associated amplitude scaling. This is a more robust
```
Good point! this is not clear, in the end I removed that final sentence as it is a little bit subjective and does not really contribute to the main aim of describing the purpose and methodology of the function.

JoeZiminski commented:
Thanks @h-mayorquin! I've addressed the points and used a bullet-point approach for the definitions I agree much clearer now, let me know what you think!

Cheers @alejoe91 I agree if these are not really in-use functions it does not make as much sense to add to the maintainability burden by adding tests for them. Do you think these would be useful enough to add as general widgets? Happy to do this, otherwise, do you think they could be moved to a personal repo? I'm conflicted, as I can see the use of having quick debugging functions in the main codebase, but if they are in the main codebase they should have some tests to check they don't become stale / unusable. But then, this adds maintainability burden 🤔

h-mayorquin commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -526,133 +587,137 @@ def fit_collision(
-#     ax.legend()
-#     ax.set_title(f"Spike {spike_index} - sample {center_spike['sample_index']}")
-#     return ax
+### Debugging ###
```
Should this be commented again?

JoeZiminski replied:
I think this is a good time to discuss, for me it is confusing to have blocks of commented code in the main codebase. It also leads to problems as when the code is wanted to be used, it cannot be because it's gone out of date (which has happened here). For me better options are (not in order of preference)  1) full inclusion as a widget + testing 2) use as debugging tool with quick tests to ensure it doesn't become unuseable 3) factoring to a private repo / deletion

alejoe91 replied:
I would keep them there uncommented. Hopefully, as we maintain the extension we can update the plotting functions as we go :)

h-mayorquin replied:
I would rather have deletion, I think this is a convenience for Sam workflow. Maybe something to discuss on an issue though. I was just trying to make the diff smaller to no distract the review.

h-mayorquin replied:
I wrote this without updating and did not see that @alejoe91 commented before me.  That is fine as well.

h-mayorquin reviewed (approved):
To me this is clearly an improvement. Some final comments but this is good to go for me.

JoeZiminski commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -384,28 +400,44 @@ def _are_unit_indices_overlapping(sparsity_mask, i, j):
+    Temporal overlap: another spike peak occurring within a specified time window
+                      around the spike peak.
+    Spatial overlap: two spikes both having signal on any shared channel (i.e.
+                     two spikes are not spatially overlapping if they have signal on
```
No I agree this is not clear, how about `if their signal is spread across two completely separate sets of channels`?

h-mayorquin replied:
This makes sense to me.

JoeZiminski commented on src/spikeinterface/postprocessing/amplitude_scalings.py:
```diff
@@ -384,28 +400,44 @@ def _are_unit_indices_overlapping(sparsity_mask, i, j):
-        An array of spikes within the added margin
+        An array of spikes, where spikes are represented by their:
+            (sample_index, channel_index, amplitude, segment_index, unit_index, in_margin)
+    spikes_within_margin: np.array
```
Nice yes, I had to break in again to realise it is a `spikes` array so I made it clear it takes the same format as `spikes`

h-mayorquin replied:
Thanks!

alejoe91 commented:
Thanks for the clarifications @JoeZiminski! I agree that this function was a bit obscure..it's very nice to see additional eyes on it!

We can keep the TODOs for a follow up, ok?

JoeZiminski commented:
Sure sounds good @alejoe91, thanks for this!
